### ISSUE TITLE
Discussion: alternative auth solutions

### ISSUE THREAD

**Issue by PeterJCLaw** [srobo/tasks#610]
Currently we have our own LDAP based user account system. This allows us to authenticate team members for at least the following services:
- IDE - online code development (also team status updates, though this is currently unused)
- IDE repo access (git push and pull with HTTP auth)
- forums
- virtual competition code submission
- self-user management (name, email, password)
- user management by team leaders (both management of existing accounts, including team assignment, and account creation)
- user management by volunteers (both via web and more direct database access via command line tool)
- media consent pre-fill download (currently unused)
- confirmation of received media consent by volunteers
- ticket download (once media consent is granted we generate a digitally signed ticket as a more easily processed proxy for the media consent form at competition reception); see also #286
- ticket validation by competition reception

Most of these services integrate directly into the LDAP server. Some use the HTTP API provided by the user management interface and others use their own LDAP facades.

It has been suggested that we could move to using OAuth2 for some or all of our auth needs. However any replacement will need to provide support for all of these services (directly or via an integrating service) and/or we will need to demonstrate why it is reasonable to remove the affected services.

**Comment by PeterJCLaw** [comment]
Personally the directly I've been looking for us to move is to have more things integrate with the HTTP API exposed by our user management interface, as this would make them agnostic of the underlying provision of the user accounts.

I can also see an argument towards introduce an OAuth2 provide on top of our existing database and then having our other services use OAuth2 against that provider. Currently though this feels like it would probably be somewhat more complex.

**Comment by prophile** [comment]
OAuth2 would be more complex but we might be able to lean more heavily on existing code.

What's the issue with LDAP?

**Comment by RealOrangeOne** [comment]
Something i'd be interested in considering is something external, like Authy. This way there's less for us to deal with and build, plus we can lean on external libraries. I don't know what hoops we'd have to jump through to to make it work for our model.

**Comment by trickeydan** [comment]
I'd also be interested in seeing us use something external and conforming to some kind of standard.

Using a standard, such as OAuth2, will allow us to pull bits of software off the shelf and quickly integrate them with our systems. OAuth2 is pretty easy to implement using libraries such as `oauthlib` as is built into many things (e.g forum software).

Now why not LDAP? LDAP is great if you're running everything off the same network and *trust* everything that you are dealing with. However, running LDAP over the internet is generally not recommended, even with LDAPS; this encourages running all of the services on one server, leading to a high degree of bus factoring. It is also difficult to configure and maintain, and the schema is not well standardised. LDAP generally requires services to handle credentials directly, which is not recommended unless absolutely necessary as it increases attack vectors against our infrastructure.

I would encourage an approach that uses something like Authy, Auth0 or [FusionAuth](https://fusionauth.io/) to replace LDAP entirely, nemesis can then be shimmed to provide support for our legacy applications before they are upgraded.

TLDR; OAuth2 is more standard and secure. Nemesis is non-standard and undocumented (afaik). Why re-invent the wheel with a custom auth standard that will likely haunt us in years to come. We can probably transition away gracefully.

**Comment by trickeydan** [comment]
Should probably add, both G Suite and GitHub support OAuth2, which would allow us to quickly check volunteer status and reduce the number of places this information is stored. Obviously we can't use this for student accounts though.

**Comment by PeterJCLaw** [comment]
> Nemesis is non-standard and undocumented (afaik).

Allow me to enlighten you: https://github.com/PeterJCLaw/nemesis/blob/master/spec.md

**Comment by trickeydan** [comment]
> Allow me to enlighten you: [PeterJCLaw/nemesis@`master`/spec.md](https://github.com/PeterJCLaw/nemesis/blob/master/spec.md?rgh-link-date=2020-07-17T21%3A19%3A29Z)

Thanks for this. I've taken a better look at Nemesis now, and it is definitely much simpler than OAuth2. I've raised some issues on the repo which we can address in the future.

As an recent example of how OAuth2 will reduce overhead for us in the future, I'd like to point towards Peter's recent work on `code-submitter`, which integrates with Nemesis. It is written using Starlette, and a custom auth backend had to be implemented. If we were using OAuth, we could have just included [`starlette-oauth2`](https://pypi.org/project/starlette-oauth2/) in the dependencies.

**Comment by PeterJCLaw** [comment]
As we have moved away from competitors having user accounts which we track/administer/have anything to do with this issue is no longer relevant. Competitors do have Discord accounts, but we don't do anything with those; the only credential interaction there is a team-level password. Tickets (and that generation of media-consent processes) no longer exist, simplifying that side of things substantially. The IDE is no longer used. The code-submitter still exists and while it has a nemesis backend, in practice we just issue credentials to teams as a whole and back it with a simple data file.

The "alternative" solution we have thus settled on is "competitors don't have accounts for us to worry about" :smile:
